### PR TITLE
Update camera plugin loading and add a check to csminit that the camera can be instantiated.

### PR DESCRIPTION
**Pull request by krlberry**
## Description
Update camera plugin loading and add a check to csminit that the camera model can be instantiated. Work completed with @scsides.

## Related Issue
CSM Camera in ISIS work.

## Motivation and Context
CSM Camera in ISIS work.

## How Has This Been Tested?
Tested locally on one file.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/base/apps/csminit/csminit.cpp [hunk 1]
```diff
@@ -199,6 +193,13 @@ namespace Isis {
     }
     cube->putGroup(infoGroup);
 
+    // TODO:
```
Comment (krlberry) [inline isis/src/base/apps/csminit/csminit.cpp]: This is our todo list for tomorrow morning!

File: isis/src/base/objs/CSMCamera/CSMCamera.cpp [hunk 2]
```diff
@@ -43,16 +43,36 @@ find files of those names at the top level of this repository. **/
 using namespace std;
 
 namespace Isis {
+
+  /**
+   * Constructor for the USGS CSM Camera Model inside ISIS.
+   */
+  CSMCamera::CSMCamera(Cube &cube, StringBlob &state) : Camera(cube) {
```
Comment (jessemapel) [inline isis/src/base/objs/CSMCamera/CSMCamera.cpp]: I do not like the idea of constructing from a BLOB. BLOBs have to do with file I/O and not in memory passing data around. I would rather this construct from the actual data: plugin name, model name, and state string.

File: isis/src/base/apps/csminit/csminit.cpp [hunk 3]
```diff
@@ -281,33 +291,63 @@ namespace Isis {
...
+    cube->deleteBlob("Table", "CameraStatistics");
+    cube->deleteBlob("Polygon", "Footprint");
+
+    // Write CSM State blob to cube
+    cube->write(csmStateBlob);
     p.WriteHistory(*cube);
-  }
 
+    try {
+      CameraFactory::Create(*cube);
```
Comment (jessemapel) [inline isis/src/base/apps/csminit/csminit.cpp:357]: Should this call the CSMCamera Cube constructor to be consistent with the previous check?

File: isis/src/base/objs/CameraFactory/CameraFactory.cpp [hunk 4]
```diff
@@ -128,39 +130,49 @@ namespace Isis {
     }
   }
 
+  
+  /**
+   * Just initialize the CSM and ISIS plugin lists.
+   */
+  void CameraFactory::Create() {
```
Comment (jessemapel) [inline isis/src/base/objs/CameraFactory/CameraFactory.cpp]: Calling Create from a factory that neither takes any data in or returns anything does not make sense to me. I would much rather this say something like initializePlugins.

Comment (jessemapel) [reply isis/src/base/objs/CameraFactory/CameraFactory.cpp]: Another option would be to make initPlugin public as it's not safe to call multiple times.

**Review by jessemapel — COMMENTED** [review]
Left a few comments. Primarily, this still needs a test that checks that it properly restores the cube on failure.

File: isis/src/base/apps/csminit/csminit.cpp (hunk 3 above)
Comment (scsides) [reply isis/src/base/apps/csminit/csminit.cpp:357]: The first test is there to see if we can create the camera, this one is to make sure we wrote all the data to the cube file correctly and can create one from the file.

File: isis/src/base/objs/CSMCamera/CSMCamera.cpp (hunk 2 above)
Comment (scsides) [reply isis/src/base/objs/CSMCamera/CSMCamera.cpp]: Agreed, blob is not best here. We'll take another look at this constructor.

File: isis/src/base/objs/CameraFactory/CameraFactory.cpp (hunk 4 above)
Comment (scsides) [reply isis/src/base/objs/CameraFactory/CameraFactory.cpp]: Maybe a little too much future thinking. This class should be a singleton and not a collection of static functions. We'll fix it.

**Review by jessemapel — APPROVED** [review]

**Pull request merged**